### ISSUE TITLE
How do I pass a transaction argument which is of a custom type defined in ink! contract?

### ISSUE THREAD

**Issue by czyczk** [polkadot-js/api#4468]
I have a complex struct defined in the ink! contract with `scale::Encode` etc properly annotated. I tried to pass in the argument directly as a Javascript object like this:

```
const obj: ComplexStruct = <-snip->; // The complex object I want to pass in as the function's argument
contract.tx[funcName]({<- gas limit, value etc ->}, obj);
```

It should encode the object directly, just like strings and numbers, but somehow the argument seemed to be converted to a JSON string and `createType()` complained like:

```
Error: createType(BlbcModelDataPlainData):: Struct: Cannot decode value "{\"metadata\":{\"resourceType\":\"plain\",<-snip->\"}" (typeof string), expected an input object, map or array
    at createTypeUnsafe (<projectName>/node_modules/@polkadot/types/create/createType.cjs:75:18)
    at TypeRegistry.createTypeUnsafe (<projectName>/node_modules/@polkadot/types/create/registry.cjs:417:45)
    at TypeRegistry.createType (<projectName>/node_modules/@polkadot/types/create/registry.cjs:409:17)
    at <projectName>/node_modules/@polkadot/api-contract/Abi/index.cjs:221:32
    at Array.map (<anonymous>)
    at Abi.value (<projectName>/node_modules/@polkadot/api-contract/Abi/index.cjs:214:137)
    at Object.toU8a (<projectName>/node_modules/@polkadot/api-contract/Abi/index.cjs:172:102)
    at <projectName>/node_modules/@polkadot/api-contract/base/Contract.cjs:137:32
    at <projectName>/node_modules/@polkadot/api/promise/decorateMethod.cjs:57:26
    at new Promise (<anonymous>)
```

Is it that I passed the argument in a wrong way or something else should be taken care of?

**Comment by VargSupercolony** [comment]
@czyczk  By no means I'm well acquainted with the low-level API, but I suspect you have to first register the type in the `api.registry`, and then call the createType. AFAIK polkadot.js does not automatically register complex types from the contract to registry.

**Comment by jacogr** [comment]
The createType that complains above should be on the registry itself, ie. if memory serves me correctly it should have the type. The error looks suspicious, aka like it didn't find an object but rather something like a string, can you show the `const obj: ComplexStruct = ...` code without the snip?

**Comment by czyczk** [comment]
I tried to print the type and the content of the object and it said something like `Object` and the content is clearly a JSON object. The object is too complex to be posted here. I was writing a minimal example to illustrate this but I'm afraid I have to reply later since after I upgraded the ink! deps to `rc-8`, I can't even get my contract instantiated! I'm currently messing up with another issue at the ink! side.
